plugins/in_windows_exporter_metrics/we_cpu.c

```diff
@@ -0,0 +1,290 @@
+
+        WE_PERFLIB_COUNTER_SPEC("clock_interrupts_total",
+                                "Total number of received and serviced " \
+                                "clock tick interrupts",
+                                "core"),
+
+        WE_PERFLIB_COUNTER_SPEC("idle_break_events_total",
+                                "Total number of time processor was woken " \
+                                "from idle",
+                                "core"),
+
+        WE_PERFLIB_GAUGE_SPEC("parkings_status",
+                              "Parking Status represents whether a " \
+                              "processor is parked or not",
+                              "core"),
+
+        WE_PERFLIB_GAUGE_SPEC("core_frequency_mhz",
+                              "Core frequency in megahertz",
+                              "core"),
+
+        WE_PERFLIB_GAUGE_SPEC("processor_performance",
+                              "Processor Performance is the average " \
+                              "performance of the processor while it is " \
+                              "executing instructions, as a percentage of" \
+                              " the nominal performance of the processor." \
+                              " On some processors, Processor Performance" \
+                              " may exceed 100%",
+                              "core"),
+
+        WE_PERFLIB_TERMINATOR_SPEC()
+    };
+
+
+int we_cpu_init(struct flb_we *ctx)
+{
+    struct we_perflib_metric_source *metric_sources;
+    int                              result;
+
+    ctx->cpu.operational = FLB_FALSE;
+
+    ctx->cpu.metrics = flb_hash_create(FLB_HASH_EVICT_NONE, 64, 128);
+
+    if (ctx->cpu.metrics == NULL) {
+        return -1;
+    }
+
+    result = we_initialize_perflib_metric_specs(ctx->cmt,
+                                                ctx->cpu.metrics,
+                                                "windows",
+                                                "cpu",
+                                                &ctx->cpu.metric_specs,
+                                                full_metric_specs);
+
+    if (result != 0) {
+        return -2;
+    }
+
+    if (ctx->windows_version >= 6.05) {
+        metric_sources = full_metric_sources;
+        ctx->cpu.query = (char *) "Processor Information";
+    }
+    else {
+        metric_sources = basic_metric_sources;
+        ctx->cpu.query = (char *) "Processor";
+    }
+
+    result = we_initialize_perflib_metric_sources(ctx->cpu.metrics,
+                                                  &ctx->cpu.metric_sources,
+                                                  metric_sources);
+
+    if (result != 0) {
+        we_deinitialize_perflib_metric_specs(ctx->cpu.metric_specs);
+
+        return -3;
+    }
+
+    ctx->cpu.operational = FLB_TRUE;
+
+    return 0;
+}
+
+int we_cpu_exit(struct flb_we *ctx)
+{
+    we_deinitialize_perflib_metric_sources(ctx->cpu.metric_sources);
```

Comment:
Is `flb_free(ctx->cpu.metric_sources)` needed ?

Reply:
That's correct, initialize makes a copy of the input array which is stored there, same applies for the specs array.